prov/tcp/src/tcpx_conn_mgr.c

```diff
@@ -178,48 +180,157 @@ static int handle_poll_list(struct poll_fd_mgr *poll_mgr)
 	return ret;
 }
 
-static void handle_connect(struct poll_fd_mgr *poll_mgr,
-			   struct poll_fd_info *poll_info)
+static int proc_establish_conn(struct poll_fd_mgr *poll_mgr,
+			       struct poll_fd_info *poll_info,
+			       struct tcpx_ep *ep,
+			       int index)
 {
-	struct tcpx_ep *ep;
-	struct fi_eq_cm_entry cm_entry;
-	struct fi_eq_err_entry err_entry;
+	struct ofi_ctrl_hdr conn_req;
 	socklen_t len;
-	int ret, status;
+	int status, ret = 0;;
 
-	assert(poll_info->fid->fclass == FI_CLASS_EP);
-	ep = container_of(poll_info->fid, struct tcpx_ep, util_ep.ep_fid.fid);
+	assert(poll_mgr->poll_fds[index].revents == POLLOUT);
 
 	len = sizeof(status);
 	ret = getsockopt(ep->conn_fd, SOL_SOCKET, SO_ERROR, &status, &len);
 	if (ret < 0 || status) {
 		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "connection failure\n");
+		ret = status? status : errno;
+		goto err;
+	}
+
+	memset(&conn_req, 0, sizeof(conn_req));
+	conn_req.version = OFI_CTRL_VERSION;
+	conn_req.type = ofi_ctrl_connreq;
+	conn_req.seg_size = htons(poll_info->cm_data_sz);
+
+	ret = ofi_send_socket(ep->conn_fd, &conn_req,
+			      sizeof(conn_req), MSG_NOSIGNAL);
+	if (ret)
+		goto err;
 
-		memset(&err_entry, 0, sizeof err_entry);
-		err_entry.fid = poll_info->fid;
-		err_entry.context = poll_info->fid->context;
-		err_entry.err = status ? status : errno;
+	if (conn_req.seg_size) {
+		ret = ofi_send_socket(ep->conn_fd, poll_info->cm_data,
+				      poll_info->cm_data_sz, MSG_NOSIGNAL);
+		if (ret)
+			goto err;
+	}
+err:
+	return ret;
```

Comment:
There's no need to use a goto just to jump to a simple return statement.  Just replace the goto's with return